packages/oauth/src/state-stores/interface.ts

```diff
@@ -1,18 +1,41 @@
 import { InstallURLOptions } from '../install-url-options';
 
-// State object structure
+/**
+ * The data structure represented by the state parameter.
+ */
 export interface StateObj {
+
+  /**
+   * The timestamp that the state object was generated.
+   */
   now: Date;
+
+  /**
+   * The passed InstallURLOptions object when generating this state parameter.
+   */
   installOptions: InstallURLOptions;
 }
 
+/**
+ * Generates state parameter value in the OAuth flow.
+ * While the state parameter value works for the CSRF protection purpose,
+ * it can transfer the given InstallURLOptions value to the Redirect URL handler
+ * (Redirect URL: the default path is "/slack/oauth_redirect")
+ */
 export interface StateStore {
-  // Returned Promise resolves for a string which can be used as an
-  // OAuth state param.
-  // TODO: Revisit design. Does installOptions need to be encoded in state if metadata is static?
+
+  /**
+   * Generates a valid state parameter value, which can be decoded as a StateObj object
+   * by the verifyStateParam() method. This value may be stored on the server-side with expiration.
+   * The InstallProvider verifies if this value is set in the installer's browser session.
+   */
   generateStateParam: (installOptions: InstallURLOptions, now: Date) => Promise<string>;
```

Comment:
Although not directly related to this change, as a bolt-js user using custom state store, I've been wondering if encoding/decoding the entire InstallURLOptions into the state is necessary.

Currently the decoded InstallURLOptions returned by `verifyStateParam()` is used for the following purposes:

* [Passing `redirect_uri` for `oauth.v2.access` method](https://github.com/slackapi/node-slack-sdk/blob/e1adad09d93991cff3a54897a6985de360faeea9/packages/oauth/src/install-provider.ts#L425)
  * Maybe we can use `req.url` instead in `handleCallback()`?
* [Assigning `metadata` to `Installation`](https://github.com/slackapi/node-slack-sdk/blob/e1adad09d93991cff3a54897a6985de360faeea9/packages/oauth/src/install-provider.ts#L498)
* [Passing entire InstallURLOptions to `CallbackOptions.success/failure`](https://github.com/slackapi/node-slack-sdk/blob/e1adad09d93991cff3a54897a6985de360faeea9/packages/oauth/src/install-provider.ts#L512)
  * Do we really need this in the callbacks when we have `Installation`?

I feel the essential is only `InstallURLOptions.metadata`.

Is there any underlying reason for the design?

That said, I understand we cannot immediately change the public API (`StateStore` and `CallbackOptions`) until next major version bump.

Reply:
@yujinakayama Thanks for the comment.
>as a bolt-js user using custom state store, I've been wondering if encoding/decoding the entire InstallURLOptions into the state is necessary.

I do understand your point here. Actually, bolt-python and bolt-java (the ones I implemented as the main developer) do not have this requirement at all. This is specific to this Node SDK package and bolt-js.

>Maybe we can use req.url in handleCallback() instead?

This is off-topic a bit but `req.url` in application code does not always return the publicly available full URL (e.g., in the case of apps behind a load balancer). Thus, knowing rediect_uri may be useful in some situations. That being said, this is not a common need. Also, if a developer would like to do this, including the URL as part of state parameter is not the only way to do so.

>I feel the essential is only InstallURLOptions.metadata.
>Is there any underlying reason for the design?

I've checked the bolt-js interface designer's design document again and found that the main intention of this design seemed to be just to transfer `metadata` to `/slack/oauth_redirect` URL (`handleCallback()` in code). This means that your question here is so reasonable. Going with InstallURLOptions here might be for flexibility to easily handle future enhancements but I cannot tell the exact reason behind that.

>That said, I understand we cannot immediately change the public API (StateStore and CallbackOptions) until next major version bump.

Yes, I'm glad to know that we're on the same page on this. In the long run, we may consider adding StateStore v2 that to make 3rd party developers' custom StateStore code much simpler. That being said, having two types of `StateStore` interface can be even more confusing (particularly for the developers that are new to Slack app development). Also, the improvement in this PR should enable developers to simply go with the default `StateStore` for the purpose of OAuth CSRF protection.

Considering all the above, my plans for the near future is to provide more flexibility to developers  by adding more callbacks in both `/slack/install` and `/slack/oauth_redirect`. The `metadata` in this SDK can be useful for some use cases but if you can set cookies, it should be more flexible and safer. Here is my proposal for it: https://github.com/slackapi/node-slack-sdk/issues/1438 If you have any comments on the design and plans, please feel free to write in there.

Applying the improvements in this PR plus having these callbacks in the next release should enable you to get rid of your own StateStore for customization (let me know if I'm missing something here).

Thanks again for your insights here!

Reply:
>Also, the improvement in this PR should enable developers to simply go with the default StateStore for the purpose of OAuth CSRF protection.

Let me correct my comment a bit. I meant that, with the improvement in this PR, the default `ClearStateStore` may become safe in the acceptable level. However, the ideal implementation would be to use sever-side database to manage active state values and access the server-side data when `verifyStateParam()` is called. In Python and Java, we provide out-of-the-box solutions for this:

* https://github.com/slackapi/python-slack-sdk/tree/v3.15.1/slack_sdk/oauth/state_store
* https://github.com/slackapi/java-slack-sdk/tree/v1.20.0/bolt/src/main/java/com/slack/api/bolt/service/builtin

node-slack-sdk, specifically `@slack/oauth`, can provide something similar in future versions. I will create a new ticket issue for it later.